### PR TITLE
Fixes issues with NULL/null/empty parameters

### LINKED ISSUE #762: Please allow us to disable param defaults
This is more of a request than an issue.

Currently, when a parameterized .Rmd is created, a default value must be coded for each param. For example, if I have the following param in the YAML...

``` yaml
other: 
      value: first
      choices: [first, second, third]
      input: select
      label: Other
      multiple: yes
```

...then I must specify a default option to go into the 'value' spot. If I leave this spot blank, it throws an error*. If I try knitting this doc with the above code, then backspacing this param widget so that no value is showing, it returns the first available value (from my choices) in the rendered doc.

It seems that RMarkdown requires a default value for every param in order to render. Is there a way I could create a parameter with no defaults (i.e., a param that passes through to the output as NULL)?

*There seems to be some NULL capability built into params in RMarkdown, but it is unpredictable (or I am using it incorrectly). For example, when I leave the 'value:' spot blank (or I make it NULL) and knit the doc with parameters, I can sometimes get an option to appear in the select widget that says "Unspecified (NULL)" and another option that says "Use a custom value." Selecting the first option _does_ pass a NULL value through to the output. Selecting the second option resets the widget to display my original choices, but throws an error during the knit. It seems like the more coder-friendly alternative would be to simply allow RMarkdown to interpret these NULL values (or blank widgets) as such, without displaying the extra options.

Thank you,

Diana

### PR DESCRIPTION
**Pull request by bearloga**
- Fixes how the list of “render params not declared in YAML” is
formatted
- Adds support for NULL/null/empty parameters
- Adds related tests

Additionally, I'm _pretty sure_ this fixes #762.

### CODE REVIEW

File: R/params.R [hunk 1]
```diff
@@ -6,7 +6,11 @@ knit_params_get <- function(input_lines, params) {
   if (packageVersion('yaml') < '2.1.14') knit_params <- mark_utf8(knit_params)
   default_params <- list()
   for (param in knit_params) {
-    default_params[[param$name]] <- param$value
+    if (is.null(param$value)) {
+      default_params[param$name] <- list(NULL)
+    } else {
+      default_params[[param$name]] <- param$value
+    }
```
Comment (yihui) [inline R/params.R]: You can use `default_params[param$name] <- list(param$value)`

Comment (bearloga) [inline R/params.R]: Thank you!

**Comment by jjallaire** [comment]
@aronatkins You are (for better or worse) the expert on the implications of NULL parameter values. What's your take here? Are there other downstream implications that need accounting for?

**Comment by aronatkins** [comment]
Also fixes #729

**Comment by bearloga** [comment]
Thank you everyone for a positive experience!

**Pull request merged**